Python: Feat/Add DEVELOPER role for openai o1

ymuichiro opened the pull request:
### Motivation and Context

Currently, the OpenAI O1 model introduces a new role "developer". However, passing the conventional "system" role results in an error. To address this issue, the following changes are proposed.

### Description

1. **Add a Method to the `ChatHistory` Class**  
   - Implement a method to handle role conversion or substitution logic for compatibility with the O1 model.

2. **Expand the `AuthorRole` Enum**  
   - Add `"developer"` as a new value to the `AuthorRole` enum.

3. **Improve Developer Experience (UX)**  
   - If the `"system"` role is mistakenly passed, the logic should internally convert it to `"developer"` for better UX.  
   - However, since all models other than O1 still use `"system"`, simply adding support for the `"developer"` role seems to be the most optimal solution at this point.

- Closes: #10032 

### Background

- The O1 model no longer supports the `"system"` role, causing errors when it is passed.  
- The introduction of the `"developer"` role must be integrated in a way that preserves compatibility with other models.

### Usage

```python
import asyncio

from semantic_kernel.connectors.ai.function_choice_behavior import FunctionChoiceBehavior
from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.azure_chat_prompt_execution_settings import (
    AzureChatPromptExecutionSettings,
)
from semantic_kernel.connectors.ai.open_ai.services.azure_chat_completion import AzureChatCompletion
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.kernel import Kernel

async def main():
    kernel = Kernel()
    service_id = "azure-openai"
    kernel.add_service(
        service=AzureChatCompletion(
            service_id=service_id,
            api_key=*********,
            endpoint=*********,
            deployment_name="o1",
            api_version="2024-12-01-preview",
        )
    )

    settings = kernel.get_prompt_execution_settings_from_service_id(service_id=service_id)

    if isinstance(settings, AzureChatPromptExecutionSettings):
        settings.function_choice_behavior = FunctionChoiceBehavior.Auto(auto_invoke=True)

    service = kernel.get_service(service_id=service_id)

    if not isinstance(service, AzureChatCompletion):
        raise Exception("Invalid Value")

    history = ChatHistory()
    history.add_developer_message("You are a helpful assistant.")
    history.add_user_message("hello")

    result = await service.get_chat_message_contents(chat_history=history, settings=settings, kernel=kernel)

    if not result:
        raise Exception("result is None")

    print(result[0].content)

if __name__ == "__main__":
    asyncio.run(main())
```

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

moonbox3 commented:
Thanks for having a look at this, @ymuichiro. I think it’d be good to make sure other o1 api features are in as well. I saw they have a reasoning_effort param, too. Are there other features we’d want to get in to make sure the o1 experience works well?

ymuichiro commented:
@moonbox3 

Thank you for checking this so quickly. Regarding o1, here are the changes I’m aware of:

※ It might be worth considering providing `OpenAIChatReasoningBase` as a separate service from `OpenAIChatCompletionBase` in the future. However, since it seems that the functionality is being gradually integrated, I believe the appropriate place to incorporate these changes for now would be in `OpenAIChatPromptExecutionSettings`.

o1: 2024-12-17
API: 2024-09-01-preview

1. In o1, the `system` role has been deprecated, and a new `developer` role has been introduced.
2. In o1, `max_tokens` has been deprecated and replaced with `max_completion_tokens`.
    • **max_completion_tokens**: int
    • The value represents the sum of reasoning tokens and completion tokens, with a recommendation to specify 25,000 or more.
3. A new property, `reasoning_effort`, has been added in o1.
    • `reasoning_effort`: enum, `low`, `medium`, `high`
    • Higher values indicate that the model will spend more time processing the request.
4. A new usage metric, `reasoning_tokens`, has been introduced in o1.
    • This represents the number of tokens used during reasoning.

```json
{
  "completion_tokens": 1843,
  "prompt_tokens": 20,
  "total_tokens": 1863,
  "completion_tokens_details": {
    "audio_tokens": null,
    "reasoning_tokens": 448
  },
  "prompt_tokens_details": {
    "audio_tokens": null,
    "cached_tokens": 0
  }
}
```

Currently Unsupported 

• Parallel tool invocation
• `temperature`, `top_p`, `presence_penalty`, `frequency_penalty`, `logprobs`, `top_logprobs`, `logit_bias`

It would be nice if these were applied to settings, usage, and author_role.

eavanvalkenburg commented:
I agree @moonbox3 bit more work is needed to make sure we fully support o1, these extra settings can just be added with None defaults, but we also need to consider the rendered prompt to ChatHistory parser, that uses a system message as default, that can be left as is, but might throw when used with o1 but might also be usefull to have a look at how we can support this as well!

moonbox3 commented:
Hi @ymuichiro thanks for the follow up details. Just getting back to the office after holidays - apologies for the slow reply. 

As I have a lot of items to go through, let me dig deeper into this today, and I can provide some more comments. As per @eavanvalkenburg's comment above, we do need to think about how the chat history will be handled now that o1 doesn't support the system message -- today one can specify the system message via the constructor, so we'd need to now handle this in a different way.

Another quick thought is: could we have `OpenAIReasoningPromptExecutionSettings`, which would contain the new properties? That way we're not trying to add too much to the chat prompt execution settings -- nothing is needing to change with the chat models. I also kind of like `OpenAIReasoning` instead of `OpenAIChatReasoning` but that might be a personal preference. 

Let me circle back soon. And thank you again for all your work on this thus far.

eavanvalkenburg commented:
I'm not sure if a seperate PES makes things easier or harder, not having too many settings is great, but we rely on the class property for the "right" PES at different points and having two possibly valid PES's means headaches for that, we could say, check the model id, if starts with `o` then use Reasoning, but that doesn't work for Azure OpenAI since there we can't reason about which model is actually behind a deployment, so I don't think that would be ideal, I think keeping it simple and add the new params with None defaults, is still easier.

eavanvalkenburg commented:
@ymuichiro thanks for your patience, me and @moonbox3 had a chat on this, the current PR is good, and please also add the two new params (max_completion_tokens and reasoning_effort) to the existing OpenAIChatPromptExecutionSettings, we will need to have a broader discussion in our team on how to handle this more broadly so for now we will just add without breaking anyone! Thanks again!

ymuichiro commented:
@moonbox3 @eavanvalkenburg 

Thank you very much for your review. I have just added a commit with additional modifications.
Regarding the changes, I thought splitting the commits would make them too granular, so I have rebased them into a single commit titled o1 Provisional Support.

eavanvalkenburg commented:
Don't worry about your commits, the pr gets squashed on merge anyway

moonbox3 commented:
> Don't worry about your commits, the pr gets squashed on merge anyway

Was just going to write this -- you beat me to it!

markwallace-microsoft commented:
<!-- Pytest Coverage Comment: python-test-coverage-report -->
<a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/README.md"><img alt="Python Test Coverage" src="https://img.shields.io/badge/Python Test Coverage-89%25-green.svg" /></a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai/prompt_execution_settings</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/semantic_kernel/connectors/ai/open_ai/prompt_execution_settings/open_ai_prompt_execution_settings.py">open_ai_prompt_execution_settings.py</a></td><td>99</td><td>1</td><td>99%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/semantic_kernel/connectors/ai/open_ai/prompt_execution_settings/open_ai_prompt_execution_settings.py#L139">139</a></td></tr><tr><td colspan="5"><b>semantic_kernel/contents</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/semantic_kernel/contents/chat_history.py">chat_history.py</a></td><td>165</td><td>2</td><td>99%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/semantic_kernel/contents/chat_history.py#L109">109</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/afb791dac79e82cedf3e7643e86dd4ce24875a4b/semantic_kernel/contents/chat_history.py#L114">114</a></td></tr><tr><td><b>TOTAL</b></td><td><b>16760</b></td><td><b>1771</b></td><td><b>89%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 2995 | 4 :zzz: | 1 :x: | 0 :fire: | 1m 17s :stopwatch: |

moonbox3 commented:
Thanks for your help on this, @ymuichiro.

moonbox3 commented on python/semantic_kernel/connectors/ai/open_ai/prompt_execution_settings/open_ai_prompt_execution_settings.py:
```diff
@@ -96,6 +96,19 @@ class OpenAIChatPromptExecutionSettings(OpenAIPromptExecutionSettings):
         dict[str, Any] | None,
         Field(description="Additional options to pass when streaming is used. Do not set this manually."),
     ] = None
+    max_completion_tokens: Annotated[
```
My only other suggestion for this (along with fixing the unit tests), is that we should add a concept sample to show how to leverage the o1 model -- and make lots of comments around what is/ what is not supported:
- developer message instead of system
- parallel tool calls not yet supported (should set PES `parallel_tool_calls = False`
- unsupported API parameters: temperature, top_p, presence_penalty, frequency_penalty, logprobs, top_logprobs, logit_bias

ymuichiro replied:
Understood. I will add a note in the comments indicating that this is a property specific to o1. Additionally, I will include examples of how to use o1 in directories like /python/samples/concepts/reasoning/**.

ymuichiro replied:
I have added samples!

ymuichiro commented:
To resolve the test error, I investigated the cause and found a challenging issue, so I am documenting it for now. The error occurs in the Azure AI Inference section. The test verifies whether all defined roles can be parsed by the Message class of Azure AI Inference.

https://github.com/microsoft/semantic-kernel/blob/2a5e51b68274c96237317435d33398d4a9315859/python/semantic_kernel/connectors/ai/azure_ai_inference/services/utils.py#L138

To pass the test, we would need to link the AuthorRole defined on the Semantic Kernel side with the ChatRole defined in the Azure AI Inference SDK. However, the Azure AI Inference SDK does not yet include the DEVELOPER role...

https://github.com/Azure/azure-sdk-for-python/blob/3fef975949b7fcb0a45fdde599902d3e39a6c0d0/sdk/ai/azure-ai-inference/azure/ai/inference/models/_enums.py#L29

As a temporary measure, we could prepare a placeholder function with a TODO: note or similar. However, this approach risks missing future updates, so I have started considering how best to address this issue ☹️

eavanvalkenburg commented:
@ymuichiro we had a chat on this one, we think it's best if the other classes raise an error for now when presented with a DEVELOPER message, that way we don't have to introduce behavior changes later on (the alternative is to map from developer to system, but then in the future if Azure AI Inference does support Developer it will change the behavior)

ymuichiro commented:
@eavanvalkenburg 
Thank you. I have implemented support for Azure AI inference.

@moonbox3 @eavanvalkenburg 
I discovered new issues and addressed them as follows:

1. 404 error occurs even when parallel_tool_call is set to False
    - I modified the code to allow None for parallel_tool_call and excluded it from the request.

2. There are unsupported properties not documented in the OpenAI documentation 
    - The following were also not supported:
        - role: tool
        - property: tool_choice

3. reasoning_effort is not supported in the current openai-python==1.57.2
    - Changed the version in uv.lock to 1.58.0

moonbox3 commented on python/samples/concepts/reasoning/simple_reasoning.py:
```diff
@@ -0,0 +1,128 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+import asyncio
```
This is great, thanks for adding the sample. Can we please update the concept sample README with this new "reasoning" category? https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/README.md

ymuichiro replied:
@moonbox3 
ok. Is it like this: https://github.com/microsoft/semantic-kernel/commit/fd3e0dc96a861331f73000009dc55d6cde5dac33#diff-f5e406ce9b1b9d01d54d2380c88086923087415cda950c90bbee37020c856a03R25

ymuichiro commented:
sorry. It seems that the uv.lock file has a conflict. Since uv is being used, I ran `uv add "openai==1.58”`, but is a different command actually used in this case?

ymuichiro commented:
@moonbox3 @eavanvalkenburg 

I resolved the conflict in uv.lock. To update the openai package to a version that supports “reasoning_effort,” I attempted to update uv.lock using the following commands. However, since these commands consistently broke uv.lock, I ended up updating it manually.

```bash
# The command was executed with the expectation that it would resolve the issue automatically.
# however, this command ended up rewriting the markers in uv.lock
uv sync --all-extras --dev --frozen
uv lock --upgrade-package "openai==1.58"
```

Note:
As part of troubleshooting, I worked within a devcontainer on GitHub Codespaces, but I was unable to resolve the issue.

moonbox3 commented:
> @moonbox3 @eavanvalkenburg
> 
> I resolved the conflict in uv.lock. To update the openai package to a version that supports “reasoning_effort,” I attempted to update uv.lock using the following commands. However, since these commands consistently broke uv.lock, I ended up updating it manually.
> 
> ```shell
> # The command was executed with the expectation that it would resolve the issue automatically.
> # however, this command ended up rewriting the markers in uv.lock
> uv sync --all-extras --dev --frozen
> uv lock --upgrade-package "openai==1.58"
> ```
> 
> Note: As part of troubleshooting, I worked within a devcontainer on GitHub Codespaces, but I was unable to resolve the issue.

Would you like some help with this? I can pull down your branch and take a look. Let me know!

ymuichiro commented:
@moonbox3 

Thank you! This time, I was able to resolve the issue manually. Therefore, no further help is needed for this PULL REQUEST.🙏

eavanvalkenburg commented:
thanks @ymuichiro you just have a readme conflict, then we can get this in!

ymuichiro commented:
@eavanvalkenburg 

Conflicts were resolved!
